File: src/main/arraylist/cpp/ExecMemo.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <chrono>
#include <stdexcept>
#include <filesystem>
#include <unistd.h>
#include "ArrayList.h"
#include <malloc_count.h>
// Compile com: 
// g++ ExecMemo.cpp ArrayList.cpp ../../../../lib/malloc_count/malloc_count.c -o ExecMemo.out -I../../../../lib/malloc_count

using namespace std;
namespace fs = std::filesystem;

const vector<string> METHOD_NAMES = {
    "remove_last",
    "add_last",
    "get_last",
    "remove_first",
    "add_first",
    "get_first",
    "remove_middle",
    "add_middle",
    "get_middle"
};

void testArrayListOperations(const vector<int>& elements, ofstream writers[], int currentLine, int totalLines) {
    int length = elements.size();
    int middle = length / 2;
    const int RUNS = 5;

    vector<vector<size_t>> memoryPeaks(METHOD_NAMES.size(), vector<size_t>(RUNS));

    ArrayList arrayList(length);
    for (int num : elements) {
        arrayList.addLast(num);
    }

    for (int run = 0; run < RUNS; run++) {
        malloc_count_reset_peak();
        arrayList.rmvLast();
        memoryPeaks[0][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.addLast(999);
        memoryPeaks[1][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.getLast();
        memoryPeaks[2][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.rmvFirst();
        memoryPeaks[3][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.addFirst(999);
        memoryPeaks[4][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.getFirst();
        memoryPeaks[5][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.rmvIndex(middle);
        memoryPeaks[6][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.addIndex(999, middle);
        memoryPeaks[7][run] = malloc_count_peak() / 1024;

        malloc_count_reset_peak();
        arrayList.get(middle);
        memoryPeaks[8][run] = malloc_count_peak() / 1024;
    }

    for (size_t method = 0; method < METHOD_NAMES.size(); method++) {
        sort(memoryPeaks[method].begin(), memoryPeaks[method].end());
        size_t medianPeak = memoryPeaks[method][RUNS / 2];
        writers[method] << "arraylist_cpp " << medianPeak << " " << length << endl;
        writers[method].flush();
    }

    cerr << "\rProcessando linha " << currentLine << "/" << totalLines << flush;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "ERRO: número de argumentos inválido!" << endl;
        return 1;
    }

    string inputFile = argv[1];
    string outputDir = "data/results/memory/";

    try {
        fs::create_directories(outputDir);
    } catch (const fs::filesystem_error& e) {
        cerr << "ERRO ao criar diretório: " << outputDir << "\n" << e.what() << endl;
        return 1;
    }

    ofstream writers[METHOD_NAMES.size()];
    for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
        string outputFile = outputDir + METHOD_NAMES[i] + ".data";
        bool exists = fs::exists(outputFile);
        bool empty = exists ? fs::file_size(outputFile) == 0 : true;

        writers[i].open(outputFile, ios::app);
        if (!writers[i]) {
            cerr << "Erro ao abrir: " << outputFile << endl;
            return 1;
        }

        if (!exists || empty) {
            writers[i] << "estrutura_linguagem memoria tamanho" << endl;
        }
    }

    try {
        ifstream reader(inputFile);
        if (!reader) throw runtime_error("Não foi possível abrir o arquivo: " + inputFile);

        int totalLines = 0;
        string line;
        while (getline(reader, line)) {
            if (!line.empty()) totalLines++;
        }
        reader.clear();
        reader.seekg(0);

        int currentLine = 0;
        while (getline(reader, line)) {
            if (line.empty()) continue;
            currentLine++;

            vector<int> elements;
            size_t pos = 0;
            while ((pos = line.find(' ')) != string::npos) {
                elements.push_back(stoi(line.substr(0, pos)));
                line.erase(0, pos + 1);
            }
            if (!line.empty()) elements.push_back(stoi(line));

            testArrayListOperations(elements, writers, currentLine, totalLines);
        }

        cerr << "\rConcluído! Processadas " << totalLines << " linhas." << endl;

        for (auto& w : writers) w.close();

    } catch (const exception& e) {
        cerr << "\nErro: " << e.what() << endl;
        for (auto& w : writers) if (w.is_open()) w.close();
        return 1;
    }

    return 0;
}

File: src/main/arraylist/python/ExecMemo.py
import sys
import os
import tracemalloc
from statistics import median
from ArrayList import ArrayList

output_dir = "data/results/memory"
os.makedirs(output_dir, exist_ok=True)

methods = ["add_first",
           "add_last",
           "add_middle",
           "remove_first",
           "remove_last",
           "remove_middle",
           "get_first",
           "get_last",
           "get_middle"]

def mem_diff(snapshot1, snapshot2):
    return sum(stat.size_diff for stat in snapshot2.compare_to(snapshot1, 'lineno'))

def write_result(method, median_mem, n):
    file_path = os.path.join(output_dir, f"{method}.data")
    is_new_file = not os.path.exists(file_path) or os.stat(file_path).st_size == 0

    with open(file_path, "a") as f:
        if is_new_file:
            f.write("estrutura_linguagem memoria tamanho\n")
        f.write(f"arraylist-python {median_mem} {n}\n")

for line_number, line in enumerate(sys.stdin, 1):
    line = line.strip()
    if not line:
        continue

    print(f"[+] Processando linha {line_number}...", file=sys.stderr)

    input_list = list(map(int, line.split()))
    n = len(input_list)
    middle = n // 2

    # Inicializa a ArrayList com os elementos de entrada
    array_list = ArrayList(input_list)
    
    mem_usage = {method: [] for method in methods}

    for _ in range(30):
        # Testa addFirst
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.addFirst(999)
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["add_first"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)  # Atualiza a referência

        # Testa addLast
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.addLast(999)
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["add_last"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)

        # Testa addIndex (middle)
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.addIndex(999, middle)
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["add_middle"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)

        # Testa rmvFirst
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.rmvFirst()
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["remove_first"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)

        # Testa rmvLast
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.rmvLast()
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["remove_last"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)

        # Testa rmvIndex (middle)
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        new_list = array_list.rmvIndex(middle)
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["remove_middle"].append(mem_diff(snapshot1, snapshot2))
        array_list = ArrayList(new_list)

        # Testa getFirst
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        array_list.getFirst()
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["get_first"].append(mem_diff(snapshot1, snapshot2))

        # Testa getLast
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        array_list.getLast()
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["get_last"].append(mem_diff(snapshot1, snapshot2))

        # Testa get (middle)
        tracemalloc.start()
        snapshot1 = tracemalloc.take_snapshot()
        array_list.get(middle)
        snapshot2 = tracemalloc.take_snapshot()
        tracemalloc.stop()
        mem_usage["get_middle"].append(mem_diff(snapshot1, snapshot2))

    for method in methods:
        write_result(method, sorted(mem_usage[method])[14], n)
